src/Collections/Sequences/MergeSort.dfy

```diff
@@ -0,0 +1,57 @@
+// RUN: %dafny /compile:0 "%s"
+
+/*******************************************************************************
+*  Copyright by the contributors to the Dafny Project
+*  SPDX-License-Identifier: MIT 
+*******************************************************************************/
+
+include "../../Relations.dfy"
+
+module Seq.MergeSort {
+
+  import opened Relations
+
+  //Splits a sequence in two, sorts the two subsequences using itself, and merge the two sorted sequences using `MergeSortedWith`
+  function method MergeSortBy<T>(a: seq<T>, compare: (T, T) -> bool): (result :seq<T>)
```

Comment:
```suggestion
  function method MergeSortBy<T>(a: seq<T>, lessEqThan: (T, T) -> bool): (result :seq<T>)
```

Reply:
It's perhaps worth noting that we could change this in the future without breaking client code. I'd maybe prefer `lessEq` or `lessThanOrEq`, to avoid the awkward reading of "equal than".

Reply:
Thanks for suggestions , I am changing `compare` to `lessThanOrEq`.

Reply:
> @MikaelMayer I just noticed you had some style requests that were lost in the shuffle. I'll respond to the ones I see that still seem relevant, but please just provide a quick comment at the top level if there are others we missed.
> 
> Re: the naming convention for predicates, e.g. "IsTotalOrder" instead of "TotalOrder", that's potentially a good idea but the latter convention is already established here in predicates like `Functions.Injective`. I'd suggest if you still feel strongly about it that you cut a repo issue for discussion, similar to #47.
> 
> Re: "lessThan" instead of "less", I don't have a strong opinion so I'd be willing to change it, personally. @prvshah51?

Could you use the attribute {:options "/functionSyntax:4"} and change all your predicate method to predicate ?

function method {:opaque} Sort<T>(s: seq<T>, compare: (T, T) -> bool) 
instead of a method

Add the following comment  to MergeSortBy
 // Splits a sequence in two, sorts the two subsequences using itself, and merge the two sorted sequences using `MergeSortedBy`

Reply:
> Could you use the attribute {:options "/functionSyntax:4"} and change all your predicate method to predicate ?
>
> function method {:opaque} Sort(s: seq, compare: (T, T) -> bool)
> instead of a method

I did that for `MergeSort.dfy`, but for `Relations.dfy` `SortedBy` was the only predicate method, and I actually made that a non-compiled `predicate` instead like the others. As per my comment on the original version in `dafny-reportgenerator`:

  // TODO: Make this a predicate-by-method so it has linear runtime instead of quadratic.
  // For now it's only used at runtime in tests.

This change doesn't need it to be compilable yet so I'd rather address that later.

> Add the following comment to MergeSortBy
> // Splits a sequence in two, sorts the two subsequences using itself, and merge the two sorted sequences using 
> MergeSortedBy

It's already there? :)

Reply:
Trying to use `{:options "/functionSyntax:4"}` ran into what looks like a bug in Dafny 3.5.0 (which this repo is using) that I can't reproduce with the latest. I've reverted it for now and we can come back to that too: https://github.com/dafny-lang/libraries/issues/55